worker/src/main/scala/org/apache/celeborn/service/deploy/worker/FetchHandler.scala

```diff
@@ -94,15 +94,168 @@ class FetchHandler(val conf: CelebornConf, val transportConf: TransportConf)
+        }
+      case PartitionType.MAP =>
+        creditStreamManager.registerStream(
+          creditStreamHandler,
+          client.getChannel,
+          initialCredit,
+          startIndex,
+          endIndex,
+          fileInfo)
+        (-1, -1)
+      case PartitionType.MAPGROUP =>
+        (-1, -1)
+    }
+  }
+
+  def handleOpenStream(
+      client: TransportClient,
+      requestId: Long,
+      msg: PbOpenStream,
+      releaseFunc: () => Unit): Unit = {
+    val (shuffleKey, fileName, startIndex, endIndex, initialCredit) =
+      (msg.getShuffleKey, msg.getFileName, msg.getStartIndex, msg.getEndIndex, msg.getInitialCredit)
+    try {
+      val (streamId, numChunks) = handleOpenStreamInternal(
+        client,
+        shuffleKey,
+        fileName,
+        startIndex,
+        endIndex,
+        initialCredit,
+        new Consumer[lang.Long] {
+          override def accept(streamId: lang.Long): Unit = {
+            val streamHandler = PbStreamHandler.newBuilder().setStreamId(streamId).build()
+            replyStreamHandler(client, requestId, streamHandler)
+          }
+        })
+      replyStreamHandler(client, requestId, fileName, startIndex, endIndex, streamId, numChunks)
+    } catch {
+      case e: IOException =>
+        handleRpcIOException(client, requestId, e)
+    } finally {
+      workerSource.stopTimer(WorkerSource.OPEN_STREAM_TIME, shuffleKey)
+      releaseFunc()
+    }
+  }
+
+  private def replyStreamHandler(
+      client: TransportClient,
+      requestId: Long,
+      fileName: String,
+      startIndex: Int,
+      endIndex: Int,
+      streamId: Long,
+      numChunks: Int,
+      legacy: Boolean = false): Unit = {
+    var streamHandler: PbStreamHandler = null
+    var legacyStreamHandler: StreamHandle = null
+    if (streamId == 0) {
+      // read from HDFS return empty stream handler
+      if (legacy) {
+        legacyStreamHandler = new StreamHandle(0, 0)
+      } else {
+        streamHandler = PbStreamHandler.newBuilder().build();
+      }
+    } else if (streamId == -1) {
+      // creditStream will handle this open stream
+    } else {
+      if (numChunks == 0)
+        logDebug(s"StreamId $streamId fileName $fileName startMapIndex" +
+          s" $startIndex endMapIndex $endIndex is empty.")
+      else logDebug(
+        s"StreamId $streamId fileName $fileName numChunks ${numChunks} " +
+          s"startMapIndex $startIndex endMapIndex $endIndex")
+      if (legacy) {
+        legacyStreamHandler = new StreamHandle(streamId, numChunks)
+      } else {
+        streamHandler = PbStreamHandler.newBuilder().setStreamId(streamId).setNumChunks(
+          numChunks).build()
+      }
+    }
+    if (streamHandler != null) {
+      replyStreamHandler(client, requestId, streamHandler)
+    }
+    if (legacyStreamHandler != null) {
+      replyLegacyStreamHandler(client, requestId, legacyStreamHandler)
+    }
+  }
+
+  private def replyStreamHandler(
```

Comment:
In my last review feedback, I meant that we only need to change the implementation of `handleOpenStream`. I don't think we should break it into so many methods here like `handleOpenStreamInternal`, `handleOpenStream`, `replyStreamHandler`, `replyStreamHandler`, `replyLegacyStreamHandler`, `handleLegacyOpenStream`, it's confusing to me.

Reply:
I don't think to create a huge method would be a good practice. Why not this form?